### PR TITLE
Add SPIFFE ID to sentry workload certs

### PR DESCRIPTION
**Pull request by yaron2**
This PR is part 1 of the epic #813 and does the following:

1. Adds strong [SPIFFE](https://spiffe.io/) based IDs to workload certs issued by Sentry
2. Sentry issued workload certs are now valid [SPIFFE SVIDs](https://github.com/spiffe/spiffe/blob/master/standards/X509-SVID.md#the-x509-spiffe-verifiable-identity-document)
3. Updates tests

Every workload cert now carries a SPIFFE ID containing the trust domain, namespace and ID of the app as a SAN extension in the x509 certificate.

This can now be used by the Dapr sidecar to validate the caller identity and perform access list authorization.

In addition, a nit cleanup was made to change the order of the identity elements from `ServiceAccount:Namespace` to `Namespace:ServiceAccount` to match better with the way it's done in Kubernetes.

### CODE REVIEW

File: pkg/sentry/csr/csr.go [hunk 1]
```diff
@@ -130,6 +137,33 @@ func GenerateCSRCertificate(csr *x509.CertificateRequest, subject string, signin
 	cert.Extensions = csr.Extensions
 	cert.BasicConstraintsValid = true
 	cert.SignatureAlgorithm = csr.SignatureAlgorithm
+
+	if identityBundle != nil {
+		spiffeID, err := identity.CreateSPIFFEID(identityBundle.TrustDomain, identityBundle.Namespace, identityBundle.ID)
+		if err != nil {
+			return nil, errors.Wrap(err, "error genrating spiffe id")
```
Comment (tcnghia) [inline pkg/sentry/csr/csr.go:144]:
```suggestion
			return nil, errors.Wrap(err, "error generating spiffe id")
```

File: pkg/sentry/identity/spiffe.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// CreateSPIFFEID returns a SPIFFE standard unique id for the given trust domain, namespace and appID.
+func CreateSPIFFEID(trustDomain, namespace, appID string) (string, error) {
+	if trustDomain == "" {
+		return "", errors.New("can't create spiffe id: trustDomain is empty")
```
Comment (tcnghia) [inline pkg/sentry/identity/spiffe.go:13]:
```suggestion
		return "", errors.New("can't create spiffe id: trust domain is empty")
```

for consistency with later errors?

**Review by tcnghia — CHANGES_REQUESTED** [review]

File: pkg/sentry/csr/csr.go [hunk 3]
```diff
@@ -130,6 +137,33 @@ func GenerateCSRCertificate(csr *x509.CertificateRequest, subject string, signin
...
+		spiffeID, err := identity.CreateSPIFFEID(identityBundle.TrustDomain, identityBundle.Namespace, identityBundle.ID)
+		if err != nil {
+			return nil, errors.Wrap(err, "error generating spiffe id")
+		}
+
+		rv := []asn1.RawValue{
+			{
+				Bytes: []byte(spiffeID),
+				Class: asn1.ClassContextSpecific,
+				Tag:   6,
```
Comment (youngbupark) [inline pkg/sentry/csr/csr.go:151]: I would use constant `asn1.TagOID` defined in encoding/asn1.

File: pkg/sentry/identity/kubernetes/validator.go [hunk 4]
```diff
@@ -54,7 +54,14 @@ func (v *validator) Validate(id, token string) error {
...
 	podNs := prts[2]
-	if id != fmt.Sprintf("%s:%s", podSa, podNs) {
+
+	if namespace != "" {
+		if podNs != namespace {
+			return errors.Errorf("%s: namespace mismatch. received namespace: %s", errPrefix, namespace)
+		}
+	}
+
+	if id != fmt.Sprintf("%s:%s", podNs, podSa) {
```
Comment (youngbupark) [inline pkg/sentry/identity/kubernetes/validator.go:64]: the previous code is `podSa:podNs`. isn't it breaking change?

**Review by youngbupark — CHANGES_REQUESTED** [review]

File: pkg/sentry/identity/kubernetes/validator.go (hunk 4 above)
Comment (yaron2) [reply pkg/sentry/identity/kubernetes/validator.go:64]: No, the sidecar injector was changed accordingly to inject this new format.

File: pkg/sentry/csr/csr.go (hunk 3 above)
Comment (yaron2) [reply pkg/sentry/csr/csr.go]: Changed.

**Review by youngbupark — APPROVED** [review]

**Comment by yaron2** [comment]
@youngbupark tests were failing due to this error:

```
Failed Platform.addApps(), cannot find dapr sidecar in pod httpmetrics-6b68845f5d-p7b6b2020/09/16 00:13:27 Running teardown...
FAIL	github.com/dapr/dapr/tests/e2e/metrics	4.029s
```

**Comment by wcs1only** [comment]
Want to see if the error is persistent.

**Comment by wcs1only** [comment]
Ok, this is intermittent. Still needs a triage to see if this was a one off, pre-existing failure, or an intermittent bug introduced by this change.

**Comment by youngbupark** [comment]
@vinayada1 SPIFFE change in sentry is merged.

**Pull request merged**